File: api/webhook.js
// pages/api/telegram.js
export const config = {
  api: { bodyParser: false },
};

import fetch from "node-fetch";

// --- В памяти ---
const sessions = {};
const feed = {};
const stats = {};
const feedbackSessions = {};



// --- Переменные окружения (обязательно установить на Vercel) ---
const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_TOKEN;
const OPENROUTER_API_KEY = process.env.OPENROUTER_API_KEY || "";
const OWNER_ID = String(process.env.MY_TELEGRAM_ID || "");

// Безопасные лимиты (Telegram: ~4096 символов в сообщении)
const TELEGRAM_SEND_MAX = 3900;

// ---- Утилиты ----
function readRawBody(req) {
  return new Promise((resolve) => {
    let data = "";
    req.on("data", (chunk) => (data += chunk));
    req.on("end", () => resolve(data));
  });
}

function chunkString(str, size = TELEGRAM_SEND_MAX) {
  const chunks = [];
  for (let i = 0; i < str.length; i += size) {
    chunks.push(str.slice(i, i + size));
  }
  return chunks;
}

function safeJson(obj) {
  try {
    return JSON.stringify(obj, null, 2);
  } catch (e) {
    return String(obj);
  }
}









// ---- Основной обработчик ----
export default async function handler(req, res) {
  if (req.method !== "POST") return res.status(200).send("OK");

  const raw = await readRawBody(req);
  let update;
  try {
    update = JSON.parse(raw.toString());
  } catch (e) {
    console.error("Bad JSON:", e);
    return res.status(400).send("Bad JSON");
  }

  console.log("📩 Получен update:", update?.update_id);

  // 1) Если это сообщение от владельца — проверим /reply
  const fromId = String(
    update?.message?.from?.id ??
      update?.edited_message?.from?.id ??
      update?.callback_query?.from?.id ??
      update?.inline_query?.from?.id ??
      ""
  );

  const isOwner = fromId && OWNER_ID && fromId === OWNER_ID;

  const msgText =
    update?.message?.text ??
    update?.edited_message?.text ??
    update?.callback_query?.data ??
    update?.inline_query?.query ??
    "";

  // Обработка команды /reply от владельца
  if (isOwner && typeof msgText === "string" && msgText.startsWith("/reply ")) {
    const parts = msgText.split(" ");
    const targetId = parts[1];
    const replyText = parts.slice(2).join(" ");
    if (!targetId || !replyText) {
      await sendMessage(OWNER_ID, "⚠ Формат: /reply <chat_id> <текст>");
    } else {
      await sendMessage(targetId, replyText);
      await sendMessage(OWNER_ID, `✅ Сообщение отправлено пользователю ${targetId}`);
    }
    // чтобы не запутывать — не пересылаем owner's /reply владельцу
    return res.status(200).send("ok");
  }

  // 2) Пересылаем владельцу полный JSON апдейта (если апдейт не от владельца)
  if (!isOwner && OWNER_ID) {
    const header = `📡 Новое событие (update_id: ${update.update_id ?? "—"})\nСодержимое апдейта (JSON):\n`;
    const body = safeJson(update);
    const payload = header + body;
    const chunks = chunkString(payload, TELEGRAM_SEND_MAX);
    for (const c of chunks) {
      // Отправляем как моно-кодный блок, используем Markdown (кодовый блок)
      // Но экранирование тройных backticks не нужно, мы просто шлем куски.
      // Если parse_mode вызывает проблемы с символами — можно убрать parse_mode или использовать "HTML".
      await sendMessage(OWNER_ID, "```json\n" + c + "\n```", null, "Markdown");
    }
  }

  // 3) Обработка игрового поведения (message, edited_message, callback_query)
  // Используем текст/данные как событие для игры.
  // Поддержаны: message.text, edited_message.text, callback_query.data
  const chatId =
    update?.message?.chat?.id ??
    update?.edited_message?.chat?.id ??
    update?.callback_query?.message?.chat?.id ??
    null;

  // Если есть callback_query — ответим на неё чтобы убрать "крутилку"
  if (update.callback_query) {
    const cqid = update.callback_query.id;
    try {
      await answerCallbackQuery(cqid);
    } catch (e) {
      // игнорируем ошибку
    }
  }

  if (chatId) {
    const chat_id_str = String(chatId);


    if (update?.message?.contact) {
      const contact = update.message.contact;
      await sendMessage(chat_id_str, `✅ Спасибо! Я получил твой номер: ${contact.phone_number}`);
      await sendMessage(
      OWNER_ID,
      `📞 Новый контакт:\nИмя: ${contact.first_name}\nТелефон: ${contact.phone_number}\nID: ${contact.user_id}`
    );
    return res.status(200).send("ok");
  }





    const text =
      update?.message?.text ??
      update?.edited_message?.text ??
      update?.callback_query?.data ??
      "";

    // запускаем игровую логику (в памяти)
    try {
      await processGameLogic(chat_id_str, String(text || ""));
    } catch (e) {
      console.error("processGameLogic error:", e);
    }
  }

  return res.status(200).send("ok");
}

// ---- sendMessage wrapper ----
async function sendMessage(chatId, text, reply_markup = null, parse_mode = "Markdown") {
  const body = {
    chat_id: String(chatId),
    text: String(text),
  };
  if (reply_markup) body.reply_markup = reply_markup;
  if (parse_mode) body.parse_mode = parse_mode;

  try {
    const r = await fetch(`https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/sendMessage`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    return r;
  } catch (e) {
    console.error("sendMessage error:", e);
    throw e;
  }
}

async function answerCallbackQuery(callback_query_id) {
  try {
    await fetch(`https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/answerCallbackQuery`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ callback_query_id }),
    });
  } catch (e) {
    console.error("answerCallbackQuery error:", e);
  }
}

// ---- Игровая логика (вся, как у тебя) ----
async function processGameLogic(chat_id, text) {
  const session = sessions[chat_id] || {};
  



  function updateStats(localChatId, game, win) {
    if (!stats[localChatId]) stats[localChatId] = {};
    if (!stats[localChatId][game]) stats[localChatId][game] = { played: 0, wins: 0 };
    stats[localChatId][game].played++;
    if (win) stats[localChatId][game].wins++;
  }

// === Запрос контакта ===
if (text === "/contact") {
  feed[chat_id]= true;
  await sendMessage(chat_id, "📱 Пожалуйста, поделитесь своим номером телефона:", {
    keyboard: [
      [{ text: "📤 Поделиться контактом", request_contact: true }],
      [{ text: "/start" }]
    ],
    resize_keyboard: true,
    one_time_keyboard: true
  });
  return;
}







  // Feedback кнопка
  if (text === "/feedback") {
    feedbackSessions[chat_id] = true;
    await sendMessage(chat_id, "📝 Пожалуйста, введите ваш комментарий одним сообщением:");
    return;
  }

  // Приём отзыва
  if (feedbackSessions[chat_id]) {
    delete feedbackSessions[chat_id];
    const { firstName, username } = sessions[chat_id] || {};
    await sendMessage(
      OWNER_ID,
      `💬 Отзыв от ${firstName || "Без имени"} (@${username || "нет"})\nID: ${chat_id}\nТекст: ${text}`
      
    );



    await sendMessage(
      OWNER_ID,
      `/reply ${chat_id}`
      
    );

    await sendMessage(chat_id, "✅ Ваш комментарий отправлен скоро с вами свяжется!");
    return;
  }



  // /start
  if (text === "/start") {

 
    sessions[chat_id] = {};
    
    await sendMessage(chat_id, `👋 Привет! Выбери тему для теста или игру:`, {
      keyboard: [
        [{ text: "История" }, { text: "Математика" }],
        [{ text: "Английский" }, { text: "Игры 🎲" }],
        [{ text: "/feedback" }, { text: "📤 Поделиться контактом", request_contact: true }]
        
      ],
      resize_keyboard: true,
    });
    return;
  }

  if (text === "📤 Поделиться контактом") {


      await sendMessage(chat_id, "Получен");
      return;
    }
  
  
  // /stats - показать статистику
  if (text === "/stats") {
    const userStats = stats[chat_id];
    if (!userStats) {
      await sendMessage(chat_id, "Ты ещё не играл ни в одну игру.");
      return;
    }

    let msg = "📊 Твоя статистика:\n\n";
    for (const game in userStats) {
      const s = userStats[game];
      msg += `• ${game}: сыграно ${s.played}, побед ${s.wins}\n`;
    }
    await sendMessage(chat_id, msg);
    return;
  }


  // Игры меню
  if (text === "Игры 🎲") {
    await sendMessage(chat_id, "Выбери игру:", {
      keyboard: [
        [{ text: "Угадай слово" }, { text: "Найди ложь" }],
        [{ text: "Продолжи историю" }, { text: "Шарада" }],
        [{ text: "/start" }, { text: "/stats" }],
      ],
      resize_keyboard: true,
    });
    return;
  }

  // Проверка ответа для тестов (История, Математика, Английский)
  if (session.correctAnswer) {
    const userAnswer = text.trim().toUpperCase();
    const correct = session.correctAnswer.toUpperCase();
    delete sessions[chat_id].correctAnswer;

    if (userAnswer === correct) {
      await sendMessage(chat_id, "✅ Правильно! Хочешь ещё вопрос?", {
        keyboard: [
          [{ text: "История" }, { text: "Математика" }],
          [{ text: "Английский" }, { text: "Игры 🎲" }],
        ],
        resize_keyboard: true,
      });
    } else {
      await sendMessage(
        chat_id,
        `❌ Неправильно. Правильный ответ: ${correct}\nПопробуешь ещё?`,
        {
          keyboard: [
            [{ text: "История" }, { text: "Математика" }],
            [{ text: "Английский" }, { text: "Игры 🎲" }],
          ],
          resize_keyboard: true,
        }
      );
    }
    return;
  }

  // Выбор темы для теста
  if (["История", "Математика", "Английский"].includes(text)) {
    const topic = text;
    const prompt = `
Задай один тестовый вопрос с 4 вариантами ответа по теме "${topic}".
Формат:
Вопрос: ...
A) ...
B) ...
C) ...
D) ...
Правильный ответ: ... (A-D)
    `.trim();

    const reply = await askGPT(prompt);
    const match = reply.match(/Правильный ответ:\s*([A-D])/i);
    const correctAnswer = match ? match[1].trim().toUpperCase() : null;

    if (!correctAnswer) {
      await sendMessage(chat_id, "⚠️ Не удалось сгенерировать вопрос. Попробуй снова.");
      return;
    }

    const questionWithoutAnswer = reply.replace(/Правильный ответ:\s*[A-D]/i, "").trim();
    sessions[chat_id] = { correctAnswer };
    await sendMessage(chat_id, `📚 Вопрос по теме *${topic}*:\n\n${questionWithoutAnswer}`, {
      parse_mode: "Markdown",
    });
    return;
  }

  // ===== Игры: Угадай слово =====
  if (text === "Угадай слово") {
    const prompt = `
Загадай одно существительное (например: тигр, самолёт, лампа и т.д.). Опиши его так, чтобы пользователь попытался угадать, что это. Не называй само слово. В конце добавь: "Загаданное слово: ..." (но это скроем от пользователя).
Формат:
Описание: ...
Загаданное слово: ...
    `.trim();

    const reply = await askGPT(prompt);
    const match = reply.match(/Загаданное слово:\s*(.+)/i);
    const hiddenWord = match ? match[1].trim().toUpperCase() : null;
    const description = reply.replace(/Загаданное слово:\s*.+/i, "").replace(/Описание:\s*/i, "").trim();

    if (!hiddenWord) {
      await sendMessage(chat_id, "⚠️ Не удалось сгенерировать описание. Попробуй ещё.");
      return;
    }

    sessions[chat_id] = { game: "Угадай слово", answer: hiddenWord };
    await sendMessage(chat_id, `🧠 Угадай слово:\n\n${description}`);
    return;
  }

  if (session.game === "Угадай слово") {
    const userGuess = (text || "").trim().toUpperCase();
    const correctAnswer = session.answer;
    delete sessions[chat_id];

    const win = userGuess === correctAnswer;
    updateStats(chat_id, "Угадай слово", win);

    let replyText = "";
    if (win) {
      replyText = "🎉 Правильно! Хочешь сыграть ещё?";
      if (stats[chat_id]["Угадай слово"].wins >= 5) {
        replyText += "\n🏅 Поздравляем! Ты получил ачивку: «Мастер угадывания»!";
      }
    } else {
      replyText = `❌ Неправильно. Было загадано: ${correctAnswer}\nПопробуешь ещё?`;
    }

    await sendMessage(chat_id, replyText, {
      keyboard: [[{ text: "Игры 🎲" }], [{ text: "/start" }]],
      resize_keyboard: true,
    });
    return;
  }

  // ===== Игры: Найди ложь =====
  if (text === "Найди ложь") {
    const prompt = `
Придумай три коротких утверждения на любые темы. Два из них должны быть правдой, одно — ложью. В конце укажи, какое из них ложь (например: "Ложь: №2").
Формат:
1. ...
2. ...
3. ...
Ложь: №...
    `.trim();

    const reply = await askGPT(prompt);
    const match = reply.match(/Ложь:\s*№?([1-3])/i);
    const falseIndex = match ? match[1] : null;

    if (!falseIndex) {
      await sendMessage(chat_id, "⚠️ Не удалось сгенерировать утверждения. Попробуй ещё.");
      return;
    }

    const statementText = reply.replace(/Ложь:\s*№?[1-3]/i, "").trim();
    sessions[chat_id] = { game: "Найди ложь", answer: falseIndex };

    await sendMessage(chat_id, `🕵️ Найди ложь:\n\n${statementText}\n\nОтвет введи цифрой (1, 2 или 3).`);
    return;
  }

  if (session.game === "Найди ложь") {
    const guess = (text || "").trim();
    const correct = session.answer;
    delete sessions[chat_id];

    const win = guess === correct;
    updateStats(chat_id, "Найди ложь", win);

    let replyText = "";
    if (win) {
      replyText = "🎉 Верно! Ты нашёл ложь!";
      if (stats[chat_id]["Найди ложь"].wins >= 5) {
        replyText += "\n🏅 Ачивка: «Ловкач»!";
      }
    } else {
      replyText = `❌ Нет, ложь была под номером ${correct}. Попробуешь ещё?`;
    }

    await sendMessage(chat_id, replyText, {
      keyboard: [[{ text: "Игры 🎲" }], [{ text: "/start" }]],
      resize_keyboard: true,
    });
    return;
  }

  // ===== Игры: Продолжи историю =====
  if (text === "Продолжи историю") {
    const prompt = `
Придумай короткое начало истории и три возможных продолжения. Варианты продолжения пронумеруй.
Формат:
Начало: ...
1. ...
2. ...
3. ...
    `.trim();

    const reply = await askGPT(prompt);
    const match = reply.match(/Начало:\s*(.+?)(?:\n|$)/i);
    const intro = match ? match[1].trim() : null;

    if (!intro) {
      await sendMessage(chat_id, "⚠️ Не удалось сгенерировать историю. Попробуй ещё.");
      return;
    }

    sessions[chat_id] = { game: "Продолжи историю" };
    await sendMessage(chat_id, `📖 Продолжи историю:\n\n${reply}\n\nВыбери номер продолжения (1, 2 или 3).`);
    return;
  }

  if (session.game === "Продолжи историю") {
    const choice = (text || "").trim();
    const win = ["1", "2", "3"].includes(choice);
    delete sessions[chat_id];
    updateStats(chat_id, "Продолжи историю", win);

    let replyText = win ? "🎉 Классное продолжение!" : "❌ Не похоже на вариант из списка.";

    if (win && stats[chat_id]["Продолжи история"].wins >= 5) {
      replyText += "\n🏅 Ачивка: «Сказочник»!";
    }

    await sendMessage(chat_id, replyText, {
      keyboard: [[{ text: "Игры 🎲" }], [{ text: "/start" }]],
      resize_keyboard: true,
    });
    return;
  }

  // ===== Игры: Шарада =====
  if (text === "Шарада") {
    const prompt = `
Придумай одну шараду (загадку), которая состоит из трех частей, каждая часть даёт подсказку, чтобы угадать слово. В конце напиши ответ.
Формат:
1) ...
2) ...
3) ...
Ответ: ...
    `.trim();

    const reply = await askGPT(prompt);
    const match = reply.match(/Ответ:\s*(.+)/i);
    const answer = match ? match[1].trim().toUpperCase() : null;

    if (!answer) {
      await sendMessage(chat_id, "⚠️ Не удалось сгенерировать шараду. Попробуй ещё.");
      return;
    }

    const riddleText = reply.replace(/Ответ:\s*.+/i, "").trim();
    sessions[chat_id] = { game: "Шарада", answer };
    await sendMessage(chat_id, `🧩 Шарада:\n\n${riddleText}\n\nНапиши свой ответ.`);
    return;
  }

  if (session.game === "Шарада") {
    const guess = (text || "").trim().toUpperCase();
    const correct = session.answer;
    delete sessions[chat_id];

    const win = guess === correct;
    updateStats(chat_id, "Шарада", win);

    let replyText = "";
    if (win) {
      replyText = "🎉 Молодец! Правильно угадал!";
      if (stats[chat_id]["Шарада"].wins >= 5) {
        replyText += "\n🏅 Ачивка: «Шарадист»!";
      }
    } else {
      replyText = `❌ Неправильно. Правильный ответ: ${correct}. Попробуешь ещё?`;
    }

    await sendMessage(chat_id, replyText, {
      keyboard: [[{ text: "Игры 🎲" }], [{ text: "/start" }]],
      resize_keyboard: true,
    });
    return;
  }

  // Если ничего не подошло
  await sendMessage(chat_id, "⚠️ Напиши /start, чтобы начать сначала или выбери команду из меню.");
}

// ---- askGPT через OpenRouter (как у тебя) ----
async function askGPT(prompt) {
  if (!OPENROUTER_API_KEY) return "Ошибка: нет OPENROUTER_API_KEY";

  try {
    const res = await fetch("https://openrouter.ai/api/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${OPENROUTER_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: "openai/gpt-3.5-turbo",
        messages: [{ role: "user", content: prompt }],
        temperature: 0.7,
      }),
    });
    const data = await res.json();
    if (!res.ok) {
      console.error("OpenRouter API error:", data);
      return "Ошибка генерации: " + (data.error?.message || "неизвестная ошибка");
    }
    return data.choices?.[0]?.message?.content || "Ошибка генерации.";
  } catch (e) {
    console.error("askGPT error:", e);
    return "Ошибка генерации.";
  }
}

File: api/webhook.py
import os
import json
import re
import httpx
from fastapi import FastAPI, Request
from fastapi.responses import PlainTextResponse

app = FastAPI()

# --- В памяти ---
sessions = {}
feed = {}
stats = {}
feedback_sessions = {}

# --- Переменные окружения ---
TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
OPENROUTER_API_KEY = os.getenv("OPENROUTER_API_KEY", "")
OWNER_ID = str(os.getenv("MY_TELEGRAM_ID", ""))

TELEGRAM_SEND_MAX = 3900

# ---- Утилиты ----
def chunk_string(s: str, size=TELEGRAM_SEND_MAX):
    return [s[i:i+size] for i in range(0, len(s), size)]

def safe_json(obj):
    try:
        return json.dumps(obj, indent=2, ensure_ascii=False)
    except Exception:
        return str(obj)

async def send_message(chat_id, text, reply_markup=None, parse_mode="Markdown"):
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
    body = {"chat_id": str(chat_id), "text": str(text)}
    if reply_markup:
        body["reply_markup"] = reply_markup
    if parse_mode:
        body["parse_mode"] = parse_mode
    async with httpx.AsyncClient() as client:
        r = await client.post(url, json=body)
        return r

async def answer_callback_query(callback_query_id):
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/answerCallbackQuery"
    async with httpx.AsyncClient() as client:
        await client.post(url, json={"callback_query_id": callback_query_id})

async def ask_gpt(prompt: str):
    if not OPENROUTER_API_KEY:
        return "Ошибка: нет OPENROUTER_API_KEY"
    try:
        async with httpx.AsyncClient() as client:
            res = await client.post(
                "https://openrouter.ai/api/v1/chat/completions",
                headers={
                    "Authorization": f"Bearer {OPENROUTER_API_KEY}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": "openai/gpt-3.5-turbo",
                    "messages": [{"role": "user", "content": prompt}],
                    "temperature": 0.7,
                },
            )
            data = res.json()
            if res.status_code != 200:
                return "Ошибка генерации: " + data.get("error", {}).get("message", "неизвестная")
            return data["choices"][0]["message"]["content"]
    except Exception as e:
        print("askGPT error:", e)
        return "Ошибка генерации."

# ---- Основной обработчик ----
@app.post("/api/telegram")
async def telegram_webhook(req: Request):
    try:
        update = await req.json()
    except Exception as e:
        print("Bad JSON:", e)
        return PlainTextResponse("Bad JSON", status_code=400)

    print("📩 Получен update:", update.get("update_id"))

    # 1) Определяем владельца
    from_id = str(
        update.get("message", {}).get("from", {}).get("id")
        or update.get("edited_message", {}).get("from", {}).get("id")
        or update.get("callback_query", {}).get("from", {}).get("id")
        or update.get("inline_query", {}).get("from", {}).get("id")
        or ""
    )
    is_owner = from_id and OWNER_ID and from_id == OWNER_ID

    msg_text = (
        update.get("message", {}).get("text")
        or update.get("edited_message", {}).get("text")
        or update.get("callback_query", {}).get("data")
        or update.get("inline_query", {}).get("query")
        or ""
    )

    # /reply для владельца
    if is_owner and isinstance(msg_text, str) and msg_text.startswith("/reply "):
        parts = msg_text.split(" ")
        target_id = parts[1] if len(parts) > 1 else None
        reply_text = " ".join(parts[2:]) if len(parts) > 2 else None
        if not target_id or not reply_text:
            await send_message(OWNER_ID, "⚠ Формат: /reply <chat_id> <текст>")
        else:
            await send_message(target_id, reply_text)
            await send_message(OWNER_ID, f"✅ Сообщение отправлено пользователю {target_id}")
        return PlainTextResponse("ok")

    # 2) Пересылка JSON апдейта владельцу
    if not is_owner and OWNER_ID:
        header = f"📡 Новое событие (update_id: {update.get('update_id','—')})\nСодержимое апдейта (JSON):\n"
        body = safe_json(update)
        payload = header + body
        for chunk in chunk_string(payload, TELEGRAM_SEND_MAX):
            await send_message(OWNER_ID, f"```json\n{chunk}\n```", parse_mode="Markdown")

    # 3) Обработка игр
    chat_id = (
        update.get("message", {}).get("chat", {}).get("id")
        or update.get("edited_message", {}).get("chat", {}).get("id")
        or update.get("callback_query", {}).get("message", {}).get("chat", {}).get("id")
    )

    if update.get("callback_query"):
        cqid = update["callback_query"]["id"]
        try:
            await answer_callback_query(cqid)
        except:
            pass

    if chat_id:
        chat_id_str = str(chat_id)
        first_name = (
            update.get("message", {}).get("from", {}).get("first_name")
            or update.get("edited_message", {}).get("from", {}).get("first_name")
            or update.get("callback_query", {}).get("from", {}).get("first_name")
            or ""
        )

        if update.get("message", {}).get("contact"):
            contact = update["message"]["contact"]
            await send_message(chat_id_str, f"✅ Спасибо! Я получил твой номер: +{contact['phone_number']}")
            await send_message(
                OWNER_ID,
                f"📞 Новый контакт:\nИмя: {contact.get('first_name')}\nТелефон: +{contact.get('phone_number')}\nID: {contact.get('user_id')}"
            )
            return PlainTextResponse("ok")

        text = (
            update.get("message", {}).get("text")
            or update.get("edited_message", {}).get("text")
            or update.get("callback_query", {}).get("data")
            or ""
        )

        try:
            await process_game_logic(chat_id_str, str(text or ""), first_name)
        except Exception as e:
            print("processGameLogic error:", e)

    return PlainTextResponse("ok")

# ---- Игровая логика ----
async def process_game_logic(chat_id, text, first_name):
    session = sessions.get(chat_id, {})

    def update_stats(local_id, game, win):
        if local_id not in stats:
            stats[local_id] = {}
        if game not in stats[local_id]:
            stats[local_id][game] = {"played": 0, "wins": 0}
        stats[local_id][game]["played"] += 1
        if win:
            stats[local_id][game]["wins"] += 1

    # /start
    if text == "/start" or text == "Назад":
        sessions[chat_id] = {"firstName": first_name}
        await send_message(
            chat_id,
            f"👋 Привет, {first_name or 'друг'}! Выбери тему для теста или игру:",
            reply_markup={
                "keyboard": [
                    [{"text": "История"}, {"text": "Математика"}],
                    [{"text": "Английский"}, {"text": "Игры 🎲"}],
                    [{"text": "/feedback"}, {"text": "📤 Поделиться контактом", "request_contact": True}],
                ],
                "resize_keyboard": True,
            },
        )
        return

    # /feedback
    if text == "/feedback":
        feedback_sessions[chat_id] = True
        await send_message(chat_id, "📝 Пожалуйста, введите ваш комментарий одним сообщением:")
        return

    if feedback_sessions.get(chat_id):
        del feedback_sessions[chat_id]
        fn = sessions.get(chat_id, {}).get("firstName")
        await send_message(
            OWNER_ID,
            f"💬 Отзыв от {fn or 'Без имени'} (ID: {chat_id})\nТекст: {text}"
        )
        await send_message(OWNER_ID, f"/reply {chat_id}")
        await send_message(chat_id, "✅ Ваш комментарий отправлен!")
        return

    # /stats
    if text == "/stats":
        user_stats = stats.get(chat_id)
        if not user_stats:
            await send_message(chat_id, "Ты ещё не играл ни в одну игру.")
            return
        msg = "📊 Твоя статистика:\n\n"
        for game, s in user_stats.items():
            msg += f"• {game}: сыграно {s['played']}, побед {s['wins']}\n"
        await send_message(chat_id, msg)
        return

    # Игры меню
    if text == "Игры 🎲":
        await send_message(chat_id, "Выбери игру:", reply_markup={
            "keyboard": [
                [{"text": "Угадай слово"}, {"text": "Найди ложь"}],
                [{"text": "Продолжи историю"}, {"text": "Шарада"}],
                [{"text": "Назад"}, {"text": "/stats"}],
            ],
            "resize_keyboard": True,
        })
        return

    # Тесты: История, Математика, Английский
    if text in ["История", "Математика", "Английский"]:
        prompt = f"""
Задай один тестовый вопрос с 4 вариантами ответа по теме \"{text}\".
Формат:
Вопрос: ...
A) ...
B) ...
C) ...
D) ...
Правильный ответ: ... (A-D)
""".strip()
        reply = await ask_gpt(prompt)
        match = re.search(r"Правильный ответ:\s*([A-D])", reply, re.I)
        correct = match.group(1).upper() if match else None
        if not correct:
            await send_message(chat_id, "⚠️ Не удалось сгенерировать вопрос. Попробуй снова.")
            return
        question = re.sub(r"Правильный ответ:\s*[A-D]", "", reply).strip()
        sessions[chat_id] = {"correctAnswer": correct}
        await send_message(chat_id, f"📚 Вопрос по теме *{text}*:\n\n{question}")
        return

    # Ответ на тест
    if "correctAnswer" in session:
        user_answer = text.strip().upper()
        correct = session["correctAnswer"]
        del sessions[chat_id]["correctAnswer"]
        if user_answer == correct:
            await send_message(chat_id, "✅ Правильно!")
        else:
            await send_message(chat_id, f"❌ Неправильно. Правильный ответ: {correct}")
        return

    # --- Мини-игры (Угадай слово / Найди ложь / Продолжи историю / Шарада) ---
    # Аналогично можно переписать из JS: ask_gpt(prompt), парсить ответ, класть в sessions и сверять

    await send_message(chat_id, "⚠️ Напиши /start, чтобы начать заново.")
